File: src/packet_pkg/scripts/serial_node.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
STM32串口通信节点（统一版本）
订阅：/cmd_vel
发布：/sensor/cmd_vel, /imu/data, /battery_voltage
"""
import os
import serial
import time
import rospy
import threading
import struct
from geometry_msgs.msg import Twist
from sensor_msgs.msg import Imu
from std_msgs.msg import Float32, Header

class JetBotSerialControl:
    def __init__(self):
        rospy.init_node('jetbot_serial_control', anonymous=True)

        port     = rospy.get_param('~port',     '/dev/ttyACM0')
        baudrate = rospy.get_param('~baudrate', 115200)

        os.system('sudo chmod 666 %s 2>/dev/null' % port)

        self.ser = None
        try:
            self.ser = serial.Serial(
                port=port, baudrate=baudrate,
                bytesize=serial.EIGHTBITS, parity=serial.PARITY_NONE,
                stopbits=serial.STOPBITS_ONE, timeout=0.1
            )
            rospy.loginfo("串口打开成功: %s", port)
        except Exception as e:
            rospy.logerr("串口打开失败: %s", e)
            rospy.signal_shutdown("串口连接失败")
            return

        self.cmd_vel_sub  = rospy.Subscriber('/cmd_vel', Twist, self.cmd_vel_callback)
        self.cmd_vel_pub  = rospy.Publisher('/sensor/cmd_vel', Twist,   queue_size=10)
        self.imu_pub      = rospy.Publisher('/imu/data',       Imu,     queue_size=10)
        self.voltage_pub  = rospy.Publisher('/battery_voltage', Float32, queue_size=10)

        self.current_cmd   = Twist()
        self.cmd_lock      = threading.Lock()
        self.last_cmd_time = time.time()
        self.rx_buffer     = bytearray()

        self.FRAME_HEADER = 0x7B
        self.FRAME_TAIL   = 0x7D
        self.FRAME_SIZE   = 24
        self.control_rate = 20
        self.timeout      = 0.5

        rospy.loginfo("等待小车初始化完成...")
        rospy.sleep(3)
        rospy.loginfo("小车初始化完成")

    # ── 工具函数 ──────────────────────────────────────────
    def float_to_int16(self, value):
        v = int(value * 1000)
        return max(-32768, min(32767, v))

    def int16_to_bytes(self, value):
        if value < 0:
            value += 65536
        return [(value >> 8) & 0xFF, value & 0xFF]

    # ── 回调 ──────────────────────────────────────────────
    def cmd_vel_callback(self, msg):
        with self.cmd_lock:
            self.current_cmd   = msg
            self.last_cmd_time = time.time()
        rospy.logdebug("收到控制: linear.x=%.3f, angular.z=%.3f",
                       msg.linear.x, msg.angular.z)

    # ── 发送 ──────────────────────────────────────────────
    def send_control_command(self, linear_x, angular_z):
        try:
            x_bytes = self.int16_to_bytes(self.float_to_int16(-linear_x))
            y_bytes = self.int16_to_bytes(0)
            z_bytes = self.int16_to_bytes(self.float_to_int16(-angular_z))

            frame = bytearray([0x7B, 0x00, 0x00,
                                x_bytes[0], x_bytes[1],
                                y_bytes[0], y_bytes[1],
                                z_bytes[0], z_bytes[1]])
            checksum = 0
            for b in frame:
                checksum ^= b
            frame.append(checksum)
            frame.append(0x7D)

            if self.ser and self.ser.is_open:
                self.ser.write(frame)
        except Exception as e:
            rospy.logerr("发送控制命令失败: %s", e)

    # ── 解析 ──────────────────────────────────────────────
    def parse_sensor_frame(self, data):
        if len(data) != self.FRAME_SIZE:
            return None
        if data[0] != self.FRAME_HEADER or data[23] != self.FRAME_TAIL:
            return None
        try:
            p = {}
            p['flag']    = data[1]
            p['vx']      = struct.unpack('>h', data[2:4])[0]   / 1000.0
            p['vy']      = struct.unpack('>h', data[4:6])[0]   / 1000.0
            p['vz']      = struct.unpack('>h', data[6:8])[0]   / 1000.0
            p['ax']      = struct.unpack('>h', data[8:10])[0]  / 1000.0
            p['ay']      = struct.unpack('>h', data[10:12])[0] / 1000.0
            p['az']      = struct.unpack('>h', data[12:14])[0] / 1000.0
            p['gx']      = struct.unpack('>h', data[14:16])[0] / 1000.0
            p['gy']      = struct.unpack('>h', data[16:18])[0] / 1000.0
            p['gz']      = struct.unpack('>h', data[18:20])[0] / 1000.0
            p['voltage'] = struct.unpack('>H', data[20:22])[0] / 1000.0
            return p
        except Exception as e:
            rospy.logerr("解析传感器数据失败: %s", e)
            return None

    def publish_sensor_data(self, data):
        twist = Twist()
        twist.linear.x  = data['vx']
        twist.linear.y  = data['vy']
        twist.angular.z = data['vz']
        self.cmd_vel_pub.publish(twist)

        imu = Imu()
        imu.header.stamp    = rospy.Time.now()
        imu.header.frame_id = "imu_link"
        imu.linear_acceleration.x = data['ax']
        imu.linear_acceleration.y = data['ay']
        imu.linear_acceleration.z = data['az']
        imu.angular_velocity.x    = data['gx']
        imu.angular_velocity.y    = data['gy']
        imu.angular_velocity.z    = data['gz']
        self.imu_pub.publish(imu)

        v = Float32()
        v.data = data['voltage']
        self.voltage_pub.publish(v)

        rospy.loginfo_throttle(2.0, "传感器: vx=%.2fm/s 电压=%.2fV",
                               data['vx'], data['voltage'])

    # ── 线程 ──────────────────────────────────────────────
    def control_loop(self):
        rate = rospy.Rate(self.control_rate)
        while not rospy.is_shutdown():
            try:
                with self.cmd_lock:
                    if time.time() - self.last_cmd_time > self.timeout:
                        self.send_control_command(0, 0)
                    else:
                        self.send_control_command(self.current_cmd.linear.x,
                                                  self.current_cmd.angular.z)
                rate.sleep()
            except Exception as e:
                rospy.logerr("控制循环错误: %s", e)
                rate.sleep()

    def receive_loop(self):
        while not rospy.is_shutdown():
            try:
                if self.ser and self.ser.is_open:
                    data = self.ser.read(self.ser.in_waiting or 1)
                    if data:
                        self.rx_buffer.extend(data)
                        while len(self.rx_buffer) >= self.FRAME_SIZE:
                            if self.rx_buffer[0] != self.FRAME_HEADER:
                                self.rx_buffer.pop(0)
                                continue
                            frame = self.rx_buffer[:self.FRAME_SIZE]
                            sensor_data = self.parse_sensor_frame(frame)
                            if sensor_data:
                                self.publish_sensor_data(sensor_data)
                            self.rx_buffer = self.rx_buffer[self.FRAME_SIZE:]
                else:
                    rospy.sleep(0.1)
            except Exception as e:
                rospy.logerr("接收循环错误: %s", e)
                rospy.sleep(0.1)

    def run(self):
        threading.Thread(target=self.control_loop, daemon=True).start()
        threading.Thread(target=self.receive_loop, daemon=True).start()
        rospy.loginfo("串口通信节点运行中...")
        rospy.spin()

    def shutdown(self):
        try:
            self.send_control_command(0, 0)
        except:
            pass
        if self.ser and self.ser.is_open:
            self.ser.close()

if __name__ == "__main__":
    node = None
    try:
        node = JetBotSerialControl()
        node.run()
    except (KeyboardInterrupt, rospy.ROSInterruptException):
        pass
    except Exception as e:
        rospy.logerr("程序错误: %s", e)
    finally:
        if node:
            node.shutdown()

File: src/packet_pkg/src/serial_node.cpp
/**
 * serial_node.cpp
 * STM32 serial communication node (C++)
 * Subscribe: /cmd_vel
 * Publish:   /sensor/cmd_vel, /imu/data, /battery_voltage,
 *            /sensor/temperature, /sensor/humidity, /sensor/smoke
 */
#include <ros/ros.h>
#include <geometry_msgs/Twist.h>
#include <sensor_msgs/Imu.h>
#include <std_msgs/Float32.h>

#include <serial/serial.h>

#include <thread>
#include <mutex>
#include <vector>
#include <cstdint>
#include <cmath>
#include <string>
#include <cstdlib>

static const uint8_t FRAME_HEADER = 0x7B;
static const uint8_t FRAME_TAIL   = 0x7D;
static const int     FRAME_SIZE   = 30;

class SerialNode
{
public:
    SerialNode() : running_(false)
    {
        ros::NodeHandle nh;
        ros::NodeHandle pnh("~");

        last_cmd_time_ = ros::Time::now();

        pnh.param<std::string>("port",          port_,              "/dev/ttyJETBOT");
        pnh.param<int>        ("baudrate",       baudrate_,          115200);
        pnh.param<double>     ("timeout",        cmd_timeout_,       0.5);
        pnh.param<int>        ("rate",           control_rate_,      20);

        cmd_vel_pub_ = nh.advertise<geometry_msgs::Twist>("/sensor/cmd_vel",     10);
        imu_pub_     = nh.advertise<sensor_msgs::Imu>    ("/imu/data",           10);
        voltage_pub_ = nh.advertise<std_msgs::Float32>   ("/battery_voltage",    10);
        temp_pub_    = nh.advertise<std_msgs::Float32>   ("/sensor/temperature", 10);
        humi_pub_    = nh.advertise<std_msgs::Float32>   ("/sensor/humidity",    10);
        smoke_pub_   = nh.advertise<std_msgs::Float32>   ("/sensor/smoke",       10);

        cmd_vel_sub_ = nh.subscribe("/cmd_vel", 10, &SerialNode::cmdVelCallback, this);

        openSerial();

        ROS_INFO("Waiting for robot initialization...");
        ros::Duration(3.0).sleep();
        ROS_INFO("Robot ready.");
    }

    ~SerialNode()
    {
        running_ = false;
        if (ctrl_thread_.joinable()) ctrl_thread_.join();
        if (recv_thread_.joinable()) recv_thread_.join();
        if (ser_.isOpen()) {
            sendControl(0.0, 0.0);
            ser_.close();
        }
    }

    void run()
    {
        running_ = true;
        ctrl_thread_ = std::thread(&SerialNode::controlLoop, this);
        recv_thread_ = std::thread(&SerialNode::receiveLoop, this);
        ROS_INFO("Serial node running. port=%s baudrate=%d", port_.c_str(), baudrate_);
        ros::spin();
        running_ = false;
    }

private:
    // ── serial ────────────────────────────────────────────
    bool openSerial()
    {
        try {
            ser_.setPort(port_);
            ser_.setBaudrate(baudrate_);
            serial::Timeout to = serial::Timeout::simpleTimeout(10);  // 10ms timeout
            ser_.setTimeout(to);
            ser_.open();
            ROS_INFO("Serial opened: %s", port_.c_str());
            return true;
        } catch (const std::exception &e) {
            ROS_ERROR("Serial open failed: %s", e.what());
            return false;
        }
    }

    // ── callback ──────────────────────────────────────────
    void cmdVelCallback(const geometry_msgs::Twist::ConstPtr &msg)
    {
        std::lock_guard<std::mutex> lock(cmd_mutex_);
        current_cmd_   = *msg;
        last_cmd_time_ = ros::Time::now();
    }

    // ── send ──────────────────────────────────────────────
    int16_t toInt16(double v)
    {
        return (int16_t)std::max(-32768.0, std::min(32767.0, v * 1000.0));
    }

    void sendControl(double linear_x, double angular_z)
    {
        if (!ser_.isOpen()) return;
        try {
            int16_t vx = toInt16(linear_x);
            int16_t vz = toInt16(angular_z);

            uint8_t frame[11];
            frame[0] = 0x7B;
            frame[1] = 0x00;
            frame[2] = 0x00;
            frame[3] = (vx >> 8) & 0xFF;
            frame[4] =  vx       & 0xFF;
            frame[5] = 0x00;
            frame[6] = 0x00;
            frame[7] = (vz >> 8) & 0xFF;
            frame[8] =  vz       & 0xFF;
            uint8_t chk = 0;
            for (int i = 0; i < 9; i++) chk ^= frame[i];
            frame[9]  = chk;
            frame[10] = 0x7D;

            ser_.write(frame, 11);
        } catch (const std::exception &e) {
            ROS_ERROR("Send control failed: %s", e.what());
        }
    }

    // ── parse ─────────────────────────────────────────────
    bool parseFrame(const std::vector<uint8_t> &data)
    {
        if ((int)data.size() != FRAME_SIZE)  return false;
        if (data[0]  != FRAME_HEADER)        return false;
        if (data[29] != FRAME_TAIL)          return false;

        uint8_t chk = 0;
        for (int i = 0; i < 26; i++) chk ^= data[i];
        if (chk != data[26]) {
            ROS_WARN_THROTTLE(1.0, "Checksum error: calc=0x%02X recv=0x%02X", chk, data[26]);
            return false;
        }

        auto s16 = [&](int i) -> int16_t  { return (int16_t) ((data[i] << 8) | data[i+1]); };
        auto u16 = [&](int i) -> uint16_t { return (uint16_t)((data[i] << 8) | data[i+1]); };

        double   vx      = s16(2)  / 1000.0;
        double   vy      = s16(4)  / 1000.0;
        double   vz      = s16(6)  / 1000.0;
        double   ax      = s16(8)  / 1000.0;
        double   ay      = s16(10) / 1000.0;
        double   az      = s16(12) / 1000.0;
        double   gx      = s16(14) / 1000.0;
        double   gy      = s16(16) / 1000.0;
        double   gz      = s16(18) / 1000.0;
        double   voltage = u16(20) / 1000.0;
        uint8_t  temp    = data[22];
        uint8_t  humi    = data[23];
        uint16_t smoke   = u16(24);

        // publish velocity
        geometry_msgs::Twist twist;
        twist.linear.x  = vx;
        twist.linear.y  = vy;
        twist.angular.z = vz;
        cmd_vel_pub_.publish(twist);

        // publish IMU
        sensor_msgs::Imu imu;
        imu.header.stamp    = ros::Time::now();
        imu.header.frame_id = "imu_link";
        imu.linear_acceleration.x = ax;
        imu.linear_acceleration.y = ay;
        imu.linear_acceleration.z = az;
        imu.angular_velocity.x    = gx;
        imu.angular_velocity.y    = gy;
        imu.angular_velocity.z    = gz;
        imu_pub_.publish(imu);

        // publish voltage
        std_msgs::Float32 v_msg;
        v_msg.data = (float)voltage;
        voltage_pub_.publish(v_msg);

        // publish temperature / humidity / smoke
        std_msgs::Float32 t_msg, h_msg, s_msg;
        t_msg.data = (float)temp;
        h_msg.data = (float)humi;
        s_msg.data = (float)smoke;
        temp_pub_.publish(t_msg);
        humi_pub_.publish(h_msg);
        smoke_pub_.publish(s_msg);

        ROS_INFO_THROTTLE(0.5,
            "Sensor: vx=%.2fm/s volt=%.2fV temp=%dC humi=%d%% smoke=%d",
            vx, voltage, temp, humi, smoke);

        return true;
    }

    // ── control loop ──────────────────────────────────────
    void controlLoop()
    {
        ros::Rate rate(control_rate_);
        while (running_ && ros::ok()) {
            {
                std::lock_guard<std::mutex> lock(cmd_mutex_);
                double dt = (ros::Time::now() - last_cmd_time_).toSec();
                if (dt > cmd_timeout_)
                    sendControl(0.0, 0.0);
                else
                    sendControl(current_cmd_.linear.x, current_cmd_.angular.z);
            }
            rate.sleep();
        }
    }

    // ── receive loop ──────────────────────────────────────
    void receiveLoop()
    {
        std::vector<uint8_t> buf;
        while (running_ && ros::ok()) {
            try {
                if (!ser_.isOpen()) {
                    ros::Duration(2.0).sleep();
                    openSerial();
                    buf.clear();
                    continue;
                }

                size_t avail = ser_.available();
                if (avail > 0) {
                    std::vector<uint8_t> tmp(avail);
                    ser_.read(tmp, avail);
                    buf.insert(buf.end(), tmp.begin(), tmp.end());
                } else {
                    ros::Duration(0.005).sleep(); // 5ms poll interval
                    continue;
                }

                while ((int)buf.size() >= FRAME_SIZE) {
                    if (buf[0] != FRAME_HEADER) {
                        buf.erase(buf.begin());
                        continue;
                    }
                    std::vector<uint8_t> frame(buf.begin(), buf.begin() + FRAME_SIZE);
                    if (parseFrame(frame))
                        buf.erase(buf.begin(), buf.begin() + FRAME_SIZE);
                    else
                        buf.erase(buf.begin());
                }

                // prevent buffer growing unbounded
                if (buf.size() > 300)
                    buf.clear();

            } catch (const serial::SerialException &e) {
                ROS_WARN("Serial exception, reconnecting: %s", e.what());
                buf.clear();
                try { ser_.close(); } catch (...) {}
                ros::Duration(2.0).sleep();
                openSerial();
            } catch (const std::exception &e) {
                ROS_ERROR("Receive loop error: %s", e.what());
                ros::Duration(0.1).sleep();
            }
        }
    }

    // ── members ───────────────────────────────────────────
    std::string port_;
    int         baudrate_;
    double      cmd_timeout_;
    int         control_rate_;

    serial::Serial ser_;

    ros::Publisher  cmd_vel_pub_, imu_pub_, voltage_pub_;
    ros::Publisher  temp_pub_, humi_pub_, smoke_pub_;
    ros::Subscriber cmd_vel_sub_;

    geometry_msgs::Twist current_cmd_;
    ros::Time            last_cmd_time_;
    std::mutex           cmd_mutex_;

    std::thread ctrl_thread_, recv_thread_;
    bool        running_;
};

int main(int argc, char **argv)
{
    ros::init(argc, argv, "jetbot_serial_control");
    SerialNode node;
    node.run();
    return 0;
}
